Columns: Avoid using CSS variables for block gap styles

ramonjd opened the pull request:
## Description

An experimental PR to make the Columns block gap behave should we remove the CSS var from the block styles. See:

- https://github.com/WordPress/gutenberg/pull/37360

In short, we're forcing a [flex layout ](https://github.com/WordPress/gutenberg/blob/3606ae587f338de1a8783232bbec0fbe2c60fbd9/packages/block-editor/src/layouts/flex.js) via `__experimentalLayout` and related properties. The rest is allowing flex `gap` to do its thing. 

### This branch

![Kapture 2021-12-16 at 14 59 57](https://user-images.githubusercontent.com/6458278/146305952-519d5a20-faa7-4fb5-903a-c1097741e7c7.gif)

### Trunk

![Kapture 2021-12-16 at 15 03 24](https://user-images.githubusercontent.com/6458278/146306224-2652ab2f-1e23-4ab3-84da-14ff1588d1db.gif)

### Changes included

* Opt in to the Layout block support for flex styling, hides the layout controls in the inspector
* Update CSS to remove references to the block gap CSS variable
* Update "tablet" viewport rules to allow the columns to grow/shrink to smooth over the removal of the hard-coded `50%` width at that breakpoint (the 50% rule had to be removed as we can't calculate the appropriate gap without a CSS variable. See: https://github.com/WordPress/gutenberg/pull/37436#issuecomment-998488516)
* Add a couple of `!important` CSS rules to smooth over differences between the Layout support's styles and the current styles for the Columns block

## How has this been tested?

* Insert a columns block.
* Try out columns blocks with a variety of column numbers and widths (e.g. be sure to test non-standard widths like (`25%`, `50%`, `25%`) and hard-coded widths like `200px`.
* Test the per-column vertical alignment options.
* Test that de-selecting the stack on mobile setting still allows forcing columns all onto the one row
* Try adjusting the block gap at the parent Columns block level
* Test that the columns render correctly at desktop, tablet, and mobile viewports
* Test with a variety of patterns and themes

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix with a potential breaking change for tablet viewport widths.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ramonjd commented:
Not sure if I captured [your idea](https://github.com/WordPress/gutenberg/pull/37360#issuecomment-995235852) correctly here @youknowriad 

I've made a start on removing the CSS var for the Columns block and things appear to work okay so far. 

I won't be around for the next two or so weeks, so feel free to destroy this PR if it's not relevant. 

cc @andrewserong

youknowriad commented:
This captures my idea yes, I kind of expected we'd be able to remove more styles form the block buy maybe there's some special cases for this block that we need to account for (responsiveness...)

I guess we should test this properly and see if everything still works as expected for the columns block (responsiveness, column widths, column count...)

jasmussen commented:
I'm partially AFK today, so I can only give some high level thoughts, hopefully they are useful.

- Longer term I'd love to see us retire across all blocks any "stack on mobile" buttons we have in favor of intrinsic behaviors. But I guess that toggle still needs to work here. 
- Same with the number of columns on the existing "tablet" breakpoint, it's a little arbitrary and ultimately might work better as an intrinsic behavior, for example by defining a minimum column width and letting the system handle the rest. But if we remove that tablet-breakpoint behavior here, we'll need to test any patterns that use columns, and see that they work reasonably well.
- Moving the columns towards the "native" flex seems like a good step regardless of the other changes. I understand this will let us still control the per-block blockgap from theme.json, is that correct?

It is probably a good idea to move forward with this one if it's necessary for #37360 to land. But it really could use good testing across existing patterns or designs that use columns.

andrewserong committed:
Initial commit

Co-authored-by: Ramon <ramonjd@users.noreply.github.com>

andrewserong committed:
Fix vertical alignment issue, try to find a compromise with tablet viewport sizes

andrewserong commented:
Thanks for taking a look @jasmussen! While Ramon isn't around, I thought I'd see if I can help progress this PR 🤞

> Longer term I'd love to see us retire across all blocks any "stack on mobile" buttons we have in favor of intrinsic behaviors. But I guess that toggle still needs to work here.

Yes, for the moment we still need it to account for the use case of columns that don't wrap, so I think at this stage our objective is to try to keep close to parity of existing functionality.

> Same with the number of columns on the existing "tablet" breakpoint, it's a little arbitrary and ultimately might work better as an intrinsic behavior, for example by defining a minimum column width and letting the system handle the rest.

Totally. I also like the points made in #37252 and in @aristath's exploration into minimum column width in #33330. For our purposes here, there was an issue with trying to keep the existing "tablet" breakpoint CSS as-is — it turns out that the `flex-basis` value of `50%` needed to use a `calc()` with the block gap CSS variable in order to be able to keep the two columns on the same row when using `wrap`. Since our objective in this PR is to remove the use of the CSS variable, I've tried the following:

In the "tablet" breakpoint, set `flex-grow` and `flex-shrink` to `1` so that we effectively let the columns and their flex-basis determine the best fit within the container. I'm hoping that something in this direction will get us closer to removing the hard-coded 50% columns, while still allowing the columns to collapse elegantly enough when they can. In the following example, with a `25%`, `50%`, `25%` set of columns, the tablet break point preserves something of the ratio, while still stretching to fill the container. In my testing the `flex-grow` also means that if any of the columns contains bigger content (e.g. full width images) then the columns stretch to the full width of the container, which _seems_ pleasing to me (or at least IMO, a little better than the forced 50% width columns).

| Desktop | Tablet | Mobile |
| --- | --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/146876707-f5b1523a-3901-4d80-ae57-bce89cc73ab9.png) | ![image](https://user-images.githubusercontent.com/14988353/146876783-3e26793a-4552-4768-838c-a4a30a7834f2.png) | ![image](https://user-images.githubusercontent.com/14988353/146876793-5bce8f1e-1e59-4b08-9e6a-3ba90823a3c5.png) |

This will need more testing, of course.

> But it really could use good testing across existing patterns or designs that use columns.

Very much this! I'm curious to see how this fares with real world content. If anyone can think of a better approach for the "tablet" breakpoint at this stage, let me know!

I've also added in a default `align-items` setting to fix the vertical alignment of columns. This can later be refactored to use more of the Layout support, but to keep the change in this PR smaller it seemed like a pragmatic way to fix it.

andrewserong commented:
I've just switched this PR out of draft. I _think_ it's ready for review and more feedback, but it could use a lot of testing with different patterns / themes, etc to see whether or not the change is disruptive, or if it needs further refining.

jasmussen commented:
Thanks for all the thought. Just to be sure I'm reading you clearly, this PR is meant to remove the tablet view for now, right? Because that's what I'm seeing in practice as well. This PR skips it:

![3 cols to 1 col](https://user-images.githubusercontent.com/1204802/146900085-336a445d-7c29-41b6-b797-25b99b7b2bc3.gif)

Trunk still has the tablet step:

![3 cols to 2 then 1](https://user-images.githubusercontent.com/1204802/146900113-acb3b1db-88c2-4f26-acff-0d1bd2192f53.gif)

To be clear, I personally think that's acceptable if we can test sufficiently any existing patterns that might rely on the tablet breakpoint. Also because this effectively fixes #37252, CC: @scruffian and @MaggieCabrera. 

For the longer term discussion, it would be really nice if we could find some way to have the columns be equal width, as the arbitrary widths aren't great. Anything we can do here, even if it needs moving towards grid, might be worth it. Was this helpful?

andrewserong commented:
> Was this helpful?

It sure was, thanks @jasmussen!

> Just to be sure I'm reading you clearly, this PR is meant to remove the tablet view for now, right? Because that's what I'm seeing in practice as well.

Yes, in practice we're effectively removing the tablet view, but still allowing wrap at that breakpoint if the content requires it, to _hopefully_ smooth over that viewport width. 🤞 We'll find out in testing with real content whether or not the CSS at that viewport meets out needs. @scruffian and @MaggieCabrera if either of you get time to test this PR with the patterns/themes you're working on, we're very happy for feedback on how to best handle things at that viewport width.

After tomorrow, I'll be AFK, too. Feel free to pick this up if anyone wants to try to get this PR in sooner, otherwise Ramon or I will be able to look at it again early in the near year.

jasmussen commented:
Thanks for the clarification! The wrapping on the tablet breakpoint could be a challenge, I imagine that's what's causing the arbitrary width columns in your GIF. As this gets more testing, we might just want to turn that off as well, remove the entire tablet behavior cleanly, then revisit when we can.

scruffian committed:
Remove tablet break

scruffian commented:
I added a commit which removes the difference between the tablet and mobile behaviour - now the columns stack on tablet too. I also tested with two themes that use columns. Here are the differences:

Stewart:
Before:
<img width="738" alt="Screenshot 2021-12-22 at 08 48 33" src="https://user-images.githubusercontent.com/275961/147069782-7e711a54-58ed-4120-b27e-73f720654d89.png">

After:
<img width="748" alt="Screenshot 2021-12-22 at 09 02 47" src="https://user-images.githubusercontent.com/275961/147069809-6543091b-df78-4ea6-9240-d186048b0576.png">


Remote:
Before:
<img width="731" alt="Screenshot 2021-12-22 at 09 22 57" src="https://user-images.githubusercontent.com/275961/147069852-6c661da8-96dd-4b48-94d6-a556bfcdb4b9.png">


After:
<img width="699" alt="Screenshot 2021-12-22 at 09 27 15" src="https://user-images.githubusercontent.com/275961/147069869-955a7bdb-f570-4af6-b22d-f1f5567ce819.png">

IMO this is an improvement, but I am concerned about people who expect the current layout on tablet which will now break...

MaggieCabrera commented:
> IMO this is an improvement, but I am concerned about people who expect the current layout on tablet which will now break...

I think both behaviors are desirable in different contexts.

I agree with @jasmussen here:

> As this gets more testing, we might just want to turn that off as well, remove the entire tablet behavior cleanly, then revisit when we can.

I think it's best that we remove the tablet styles that so far themes would have to override and let themes add it if they need it while we look for a better way of handling this kind of situation.

jasmussen commented:
The price of removing the break is we need to thoroughly test any patterns, including older ones, to see if results are mostly acceptable.

andrewserong commented:
Thanks everyone (and @scruffian for removing the tablet break point!), I agree, I think that's a better path forward. I'm wrapping up for the year shortly, so I'll leave this PR for everyone else to test (and feel free to merge if it winds up looking viable, of course). Happy to help test / follow up again in the new year.

ramonjd committed:
Hide the layout controls so that we can reconcile the discrepancies between layout support styles and columns stack on mobile styles. The layout controls are not shown on trunk, and there is some confusion between multiple lines wrap (layout) vs stack on mobile (columns) options.

In order to override the layout support flex-wrap styles, we do need `!important` on our wrap/nowrap declarations.

ramonjd committed:
Update style.scss

Reverting break-small change

stacimc reviewed (commented):
Removing the tablet breakpoint seems like a good path forward. +1 for the removal of the layout controls for now -- the `Allow to wrap to multiple lines` option conflicting with `Stack on mobile` felt confusing to me.

I've tested columns of varying widths and custom widths, alignments, vertical alignments, and `stack on mobile`, and it's working great for me 😄 

I've also tested the following patterns:
* Social links with a shared background color
* Three columns of text
* Three columns with images and text
* Three columns with offset images
* Two columns of text and title
* Two columns of text with offset heading
* Image on solid color with description
* Pricing table
* Contact Information
* Overlapping images
* Overlapping images and text

So far everything looks reasonable to me, except that patterns with overlapping columns (`Overlapping images and text`, for example) have some odd spacing on TT1-blocks:

| Trunk | Branch |
| --- | --- |
| <img width="590" alt="Screen Shot 2022-01-03 at 4 39 29 PM" src="https://user-images.githubusercontent.com/63313398/147997695-772c3f3e-1630-439d-94a2-ea31d7f8bac1.png"> | <img width="750" alt="Screen Shot 2022-01-03 at 4 40 04 PM" src="https://user-images.githubusercontent.com/63313398/147997709-239d0104-02fa-49cc-909c-704ec10837fa.png"> | 

Also the columns seem to collapse in `Two columns of text with offset heading` on Twenty Twenty, so the `<hr>` overlaps the paragraph:

| Trunk | Branch |
| --- | --- |
| <img width="568" alt="Screen Shot 2022-01-03 at 5 31 23 PM" src="https://user-images.githubusercontent.com/63313398/147997932-a1dee3c6-81bd-4c61-b432-4821f567162f.png"> | <img width="736" alt="Screen Shot 2022-01-03 at 5 15 03 PM" src="https://user-images.githubusercontent.com/63313398/147997945-adc2a646-a1e8-4975-8ef0-a9859783e7e1.png"> |

ramonjd commented:
Thanks for testing so thoroughly @stacimc 

> So far everything looks reasonable to me, except that patterns with overlapping columns (Overlapping images and text, for example) have some odd spacing on TT1-blocks:

This looks like a theme-specific phenomenon. 

It seems as if [TT1 is applying the top and left offset](https://themes.trac.wordpress.org/browser/tt1-blocks/0.4.8/assets/css/blocks.css#L18) for a tablet-like breakpoint (min-width 652px) 

```css
.is-style-twentytwentyone-columns-overlap .wp-block-column:nth-child(2n) {
    margin-left: calc(-2 * var(--wp--custom--spacing--horizontal));
    margin-top: calc(2.5 * var(--wp--custom--spacing--horizontal));
    z-index: 2;
}
```

> Also the columns seem to collapse in Two columns of text with offset heading on Twenty Twenty, so the <hr> overlaps the paragraph:

Looks like this is also very theme-specific. TwentyTwenty places a few negative top margins on Columns Blocks at various breakpoints, **and** when a wide or full alignment is active. 

[This one](https://themes.trac.wordpress.org/browser/twentytwenty/1.8/style.css#L5235) (min-width 700px) appears to be related to the effects we're seeing

```css
.wp-block-columns.alignwide + .wp-block-columns.alignwide,
.wp-block-columns.alignfull + .wp-block-columns.alignfull {
    margin-top: -6rem;
}
```

Given the specificity to the theme, and the unique effect that they're aiming for I wouldn't catalogue them as blockers for this PR.

I don't see these effects on these patterns at the same breakpoint when changing theme for example.

| Twenty Twenty  | Twenty Twenty Two |
| ------------- | ------------- |
| <img width="300" alt="Screen Shot 2022-01-04 at 1 19 43 pm" src="https://user-images.githubusercontent.com/6458278/148001223-66188ba1-1574-4b77-85fa-88033cae454b.png">  | <img width="300" alt="Screen Shot 2022-01-04 at 1 19 33 pm" src="https://user-images.githubusercontent.com/6458278/148001231-d65aa735-8676-49f6-8a20-c2c33acfa489.png">  |


I'd rely on @scruffian's advice, but seeing as they're very specific to the themes, perhaps we could address them in the themes themselves?

ramonjd commented:
@youknowriad 👋 

We've made some final changes and have been testing this PR today.

Aside from some theme-specific issues, things are looking pretty good in my opinion, whatever that's worth. I've followed @stacimc's [test plans](https://github.com/WordPress/gutenberg/pull/37436#pullrequestreview-843144045). 

There are probably, as usual, some unknown unknowns, but I haven't managed to come across any blaring sirens in my testing.

If you get time, it would be great to get your opinion on whether it's okay to go ahead and, more importantly, backport this one.

aaronrobertshaw reviewed (commented):
Nice work 👍 

This is testing pretty well for me. Overall my impression is that the spacing of columns is improved in a few themes without any major breakage. I did encounter the theme-specific issues others have already noted.

> -   Try adjusting the block gap at the parent Columns block level

I have one small issue/question. Should this be opting into the `blockGap` support for the Columns block? The GIF in the PR description along with the quoted test instruction suggest it should be.

If we are, would we be removing the margin block support or keeping both despite the overlap?

---

✅ Tried columns blocks with varying numbers of columns, widths and stacked on/off
✅ Tested all the per-column vertical alignments
✅ Confirmed responsive behaviours for tablet and mobile breakpoints
✅ Tried adjusting block gap at columns block level (after tweaking block.json to opt-in) 
✅ Tested with various themes and patterns

Test themes:
- TwentyTwentyTwo
- TwentyTwentyOne
- TT1 Blocks
- TwentyTwenty
- TwentyTwenty Blocks
- TwentyNineteen
- Hever

Test patterns:
- Three columns of text
- Three columns with images and text
- Pricing table
- Two columns of text and title
- Two columns of text with offset heading
- Overlapping images
- Overlapping images and text

ramonjd commented:
Thanks for the thorough testing @aaronrobertshaw 

> I have one small issue/question. Should this be opting into the blockGap support for the Columns block? The GIF in the PR description along with the quoted test instruction suggest it should be.

Looks like blockGap support was deactivated as part of removing the block gap style CSS variable in **Avoid using CSS variables for block gap styles #37360**  in [this commit](https://github.com/WordPress/gutenberg/commit/3f43af06db0b017bc09124f1ca99ed89d16cb471#diff-47a3718f24a77fdd49608c208cf5cabcf46806b869dbc532d5097fd38f5831f9L27). 

I didn't get around to updating the test description today. Will get onto that now. Sorry for the confusion!

> If we are, would we be removing the margin block support or keeping both despite the overlap?

That's a great question. In my limited view, I don't see why we couldn't keep the axial top/bottom margin. It seems to be uniform in it's application.

![Jan-04-2022 19-00-34](https://user-images.githubusercontent.com/6458278/148027629-4ac9beef-4aa8-4986-8266-80ee9048aa04.gif)

Others might have a different opinion.

Let's see if @youknowriad thinks this PR is okay to approve and backport this PR today 🎉

aaronrobertshaw commented:
> That's a great question. In my limited view, I don't see why we couldn't keep the axial top/bottom margin. It seems to be uniform in it's application.

My concern was for mobile/tablet and once the columns start stacking. Would we not end up with both the block gap and the margin being applied? As a result of Flex not having collapsing margins.

youknowriad commented:
This PR looks good to me. Should we bring back the blockGap support as well (it's removed on trunk)

ramonjd commented:
> Should we bring back the blockGap support as well (it's removed on trunk)

I've been testing this. It looks pretty stable and work as expected (that is, no real unexpected changes since I last tested in December 2021)

![Jan-04-2022 19-11-27](https://user-images.githubusercontent.com/6458278/148028844-df2e66b1-4b7f-487a-9ffe-bb3bd8c64880.gif)

I'll re-add it.

> Would we not end up with both the block gap and the margin being applied? 

Thanks for reminding me about the lack of margin-collapsing for flexbox. 

In this PR there is overlap of margins between columns. I've had a play around, and to me it seems okay in the editor as I can overcompensate to get the effect I want using a combination of axial margins and gap. 

This PR locks the columns layout into flex for now via `__experimentalLayout` support. 

Is there something else we should be testing? I recall you mentioned the possibility of [regressions in blocks](https://github.com/WordPress/gutenberg/pull/32909#issuecomment-963674509) already out there?

ramonjd committed:
Reinstate blockGap support for the columns block.

aaronrobertshaw commented:
> In this PR there is overlap of margins between columns. I've had a play around, and to me it seems okay in the editor as I can overcompensate to get the effect I want using a combination of axial margins and gap.

The column stacking part of the equation isn't an issue. I confused the `column/block.json` with the `columns/block.json` earlier and thought individual columns also had margin block support. They don't so that spacing is solely dependent on the block gap. 

> Is there something else we should be testing? I recall you mentioned the possibility of regressions in blocks already out there?

From memory, that comment was related to the content inside a column. Column blocks out in the wild potentially have floated contents so forcing them to display as flex containers could break their layout. This PR is only directly related to the parent Columns block so shouldn't be an issue.

Sounds like this is ready to go. 🚢

youknowriad commented:
I was thinking about this and 5.9. 5.8 doesn't have gap support at all in columns meaning if we don't land this in 5.9, it won't be considered a regression like in the Gutenberg plugin where the block used to have block gap support. 

So given potential impact of this PR, I was thinking that maybe it's fine to just land it in Gutenberg plugin for now? Thoughts?

ramonjd commented:
>  I was thinking that maybe it's fine to just land it in Gutenberg plugin for now? Thoughts?

Thanks for your guidance on this one. 

Landing in the plugin sounds like a safer bet. I know there's some pressure in relation to 5.9 and it would help to have the breathing space to monitor/debug/garner feedback just in case we're breaking unknowns.

umhan35 commented:
Hi all. My site is ugly now due to this breaking-change PR. There were 2em of gap between columns but now it's 0.5em. 5.8 doesn't have a block gap support yet so can't change in the editor. Is there a workaround for 5.8? I am rolling back to 12.3.2 but it's not a long-term solution. Thanks!

ramonjd commented:
> Hi all. My site is ugly now due to this breaking-change PR. There were 2em of gap between columns but now it's 0.5em. 5.8 doesn't have a block gap support yet so can't change in the editor. Is there a workaround for 5.8? I am rolling back to 12.3.2 but it's not a long-term solution. Thanks!

Hi @umhan35 

Sorry to hear that! Thanks for commenting here. 

The CSS var was creating a bunch of conflicts, creating more problems than it solved. This PR was a result of [limiting it to layout-based blocks](https://github.com/WordPress/gutenberg/pull/37360).

Are you able to edit your site's theme.json (assuming you have a block-based theme like TT1 Blocks )?

You could try enabling blockGap support for the columns block. 


So, `settings.spacing.blockGap: true` and `settings.blocks['core/columns'].spacing.blockGap:true`

```json
...
    "settings": {
          ...
                "spacing": {
                      "blockGap":true,
                         ...
                   },
                  ...
		"blocks": {
                        ...
			"core/columns": {
				"spacing": {
					"blockGap": true
				}
			}
		}
   }
```

and then head to the editor and adjust the block gap setting for that particular block:

![Jan-20-2022 17-51-04](https://user-images.githubusercontent.com/6458278/150288221-06505f33-098b-4e63-b502-d480931a7bb9.gif)

If none of these options are available to you then I think you've done the right thing in rolling back for now. 

WordPress 5.9 will be released early this year so 🤞 this will fix this particular issue for you.

Hope that helps for now. Cheers!
